api/v1alpha1/stage_types.go

```diff
@@ -303,13 +303,13 @@ type HelmImageUpdate struct {
 // HelmChartDependencyUpdate describes how a specific Helm chart that is used
 // as a subchart of an umbrella chart can be updated.
 type HelmChartDependencyUpdate struct {
-	// RegistryURL along with Name identify a subchart of the umbrella chart at
+	// Repository along with Name identify a subchart of the umbrella chart at
 	// ChartPath whose version should be updated.
 	//
 	//+kubebuilder:validation:MinLength=1
 	//+kubebuilder:validation:Pattern=`^(((https?)|(oci))://)([\w\d\.\-]+)(:[\d]+)?(/.*)*$`
-	RegistryURL string `json:"registryURL"`
-	// Name along with RegistryURL identify a subchart of the umbrella chart at
+	Repository string `json:"repository"`
```

Comment:
In the PR description, you gave these examples:
* oci://ghcr.io + akuity/kargo-charts/kargo
* oci://ghcr.io/akuity + kargo-charts/kargo
* oci://ghcr.io/akuity/kargo-charts + kargo

Does that imply that `Name` is allowed to have slashes? And if so, is that equally allowed in Helm Chart.yaml?

Reply:
Technically, no... chart names cannot have slashes in them.

What a lot of this PR comes down to is that _when using OCI repos_, people accustomed to classic chart repos are apt to do something like this in their Warehouse:

* Registry: ghcr.io/akuity/kargo-charts
* Chart: kargo

Before this PR, Kargo would choke on the above. `ghcr.io/akuity/kargo-charts` _is not a registry_. `ghcr.io` is the registry.

People more familiar with OCI repos are apt to do define their Warehouse correctly:

* Registry: ghcr.io
* Chart: akuity/kargo-charts

This _will_ work. The problem is registry + chart is not the terminology used either in a `Chart.yaml` file's dependencies section, nor is it the terminology used by Argo CD.

So if your Application looks like this:

* Repository: ghcr.io/akuity/kargo-charts
* Chart: kargo

Then, prior to this PR, these don't appear to reference the same chart as the Warehouse. ☹️ Impedance.

People more accustomed to classic chart repos will also be perplexed as to the correct use of the "registry" field in a Warehouse.

To get everyone on the same page, and in a case of "if you can't beat 'em, join 'em," I'm changing "registry" in Warehouse and Freight to "repository".

Now people familiar with OCI will be the ones to get confused writing their Warehouse:

* Repository: ghcr.io/akuity/kargo-charts/kargo <---- this _is_ a repo
* Chart: What do I put here???

They are apt to try:

* Repository: ghcr.io <---- this _is_ a registry
* Chart: akuity/kargo-charts/kargo

And once again, this won't match what's in the Application, which would look like this:

* Repository: ghcr.io/akuity/kargo-charts <---- this _isn't_ a repo, but this is how Argo CD wants you to do it
* Chart: kargo

So the other thing this PR does is always joins "repo" and "chart" together with a slash, on both sides of a comparison. This makes matching more resilient to what is in each field, as long as, when glued together, they look the same.

This is all one big application of Postel's principle is all it is.

Reply:
First, it's great that we're migrating from the "registry" terminology in favor of repository, so this PR is very welcome. 

> And once again, this won't match what's in the Application, which would look like this:
> Repository: ghcr.io/akuity/kargo-charts <---- this isn't a repo, but this is how Argo CD wants you to do it
> Chart: kargo

I believe Argo CD is this way because it was following what [Helm expects in Chart.yaml](https://helm.sh/docs/topics/registries/#specifying-dependencies). I think it is the right move for Kargo to also follow this convetion. 

My only suggestion is to disallow slashes in the name, e.g.:
* oci://ghcr.io + akuity/kargo-charts/kargo
* oci://ghcr.io/akuity + kargo-charts/kargo

Since having paths in chart names is unconventional (AFAIK). I believe we can easily enforce this with a CRD validation regexp that only allows [letters, numbers, and dashes](https://helm.sh/docs/chart_best_practices/conventions/#chart-names).

> Note that, technically, oci://ghcr.io/akuity/kargo-charts is not a repository. oci://ghcr.io/akuity/kargo-charts/kargo is the repository.

Be that as that may, I'm probably not alone in considering `oci://ghcr.io/akuity/kargo-charts` as the repository, but that could just be my ignorance caused by helm's terminology polluting my understanding.

Lastly, would all confusion be avoided if there was a *_single_* field name, `repositoryURL` (as opposed to registryURL + name) where the value would be something like `oci://ghcr.io/akuity/kargo-charts/kargo`. And would a single `repositoryURL` even be accurate for that purpose? Genuine question, for my own edification.

Reply:
> You could put repo URL + name in that field and it would work, but I still feel this may confuse some people -- the classic crowd more than the OCI crowd...

Yea... you're probably right. This point 

> If you'd be ok with chartURL in the Warehouse and Freight, I also would be happy with that,

Actually, my preference out of all of this is to have the most conventional / least surprising behavior. Which at this point still is repository + name, as opposed to chartURL.

Last question, and not trying to gate the PR, but would/should we allow Name to be empty as a fourth option?

* oci://ghcr.io + akuity/kargo-charts/kargo
* oci://ghcr.io/akuity + kargo-charts/kargo
* oci://ghcr.io/akuity/kargo-charts + kargo
* oci://ghcr.io/akuity/kargo-charts/kargo + ''

Then an OCI user could specify `repository: oci://ghcr.io/akuity/kargo-charts/kargo`, omit name field entirely, and everything would be technically accurate, based on everything you just explained to me. This seems like win/win in terms of accuracy and behavior.

NOTE: in order to support this, the code would have to ensure `/` would not be appended when name is omitted, so `fmt.Sprintf("%s/%s", repository, name)` would need to be replaced with something like `path.Join(repository name)`.

Reply:
> would/should we allow Name to be empty as a fourth option

> Then an OCI user could specify repository: oci://ghcr.io/akuity/kargo-charts/kargo, omit name field entirely, and everything would be technically accurate

@jessesuen  🙌  That is a fantastic idea.

The existing validation webhook can do complex validations. We can actually require name to be empty if repo starts with oci:// and non-empty otherwise.